### PR TITLE
Adds closure annotations to aggregators.

### PR DESCRIPTION
**Pull request by paullewis**
@brendankenny @samccone Did I annotate right?

### CODE REVIEW

**Comment by samccone** [comment]
defer to @brendankenny on this one

File: aggregators/can-load-offline/index.js [hunk 1]
```diff
@@ -19,14 +19,25 @@
 
 const Aggregate = require('../aggregate');
 
+/**
+ * @type {string}
+ */
+const serviceWorker = require('../../audits/offline/service-worker').name;
```
Comment (brendankenny) [inline aggregators/can-load-offline/index.js]: hmm, do these not get figured out automatically? Closure should be able to figure out the type itself

File: closure/typedefs/Aggregation.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/** @type {!Array<!Object>} */
+Aggregation.prototype.subItems;
+
+/**
+ * @struct
+ * @record
+ */
+function Criterion() {}
+
+/** @type {?} */
```
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: right now this is only `@type {(boolean|number|undefined)}

File: aggregators/can-load-offline/index.js [hunk 3]
```diff
@@ -19,14 +19,25 @@
...
+  /**
+   * @override
+   */
   static get name() {
     return 'Works Offline';
   }
 
+  /**
+   * @override
+   * @return
```
Comment (brendankenny) [inline aggregators/can-load-offline/index.js]: `@return {!Criteria}`?

File: aggregators/aggregate.js [hunk 4]
```diff
@@ -19,19 +19,38 @@
...
+   * @return {!Array<!AuditResult>}
+   */
   static _filterResultsByAuditNames(results, expected) {
     const expectedNames = Object.keys(expected);
     return results.filter(r => expectedNames.indexOf(r.name) !== -1);
   }
 
+  /**
+   * @private
+   * @param  {!Criteria} expected
```
Comment (brendankenny) [inline aggregators/aggregate.js]: ugh, I really dislike the aligned braces in JSDocs thing. We should pick a style here :)

File: aggregators/aggregate.js [hunk 5]
```diff
@@ -89,6 +120,34 @@ class Aggregate {
...
+  static _convertBooleanToWeight(resultValue, expectedValue, weight) {
+    return (resultValue === expectedValue) ? weight : 0;
+  }
+
+  /**
+   * Compares the set of audit results to the expected values.
+   * @param  {!Array<!AuditResult>} results The audit results.
+   * @param  {!Criteria} expected The aggregation's expected values and weighting.
+   * @return {Aggregation} The aggregation score.
+   */
```
Comment (brendankenny) [inline aggregators/aggregate.js]: github won't let me comment below this, but `aggregate` also needs annotations?

File: closure/typedefs/Aggregation.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+Aggregation.prototype.overall;
+
+/** @type {!Array<!Object>} */
+Aggregation.prototype.subItems;
+
+/**
+ * @struct
+ * @record
+ */
+function Criterion() {}
```
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: Can we call these something like `AggregationCriteria`/`AggregationCriterion`?

File: closure/typedefs/Aggregation.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+/** @type {?} */
+Criterion.prototype.value;
+
+/** @type {number} */
+Criterion.prototype.weight;
+
+/**
+ * @type {Object<string, Criterion>}
```
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: `{!Object<string, !Criterion>}`

File: closure/typedefs/Aggregation.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * @struct
+ * @record
+ */
+function Aggregation() {}
+
+/** @type {number} */
+Aggregation.prototype.overall;
+
+/** @type {!Array<!Object>} */
```
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: `{!Array<!AuditResult>}`, I believe

File: aggregators/can-load-offline/index.js (hunk 1 above)
Comment (brendankenny) [inline aggregators/can-load-offline/index.js]: oh right, this may just be from what you mentioned about requires needing to be at the top level. It would be interesting to see if that's because it uses the same underlying system as ES2015 modules

**Comment by brendankenny** [comment]
reviewed. This is a good start. As we add types we should look out for ways to simplify this process to make sure the compiler is working for us and not the other way around.

Some of this current PR should be throwing errors and isn't, unfortunately, which is partly due to the optional/gradual typing approach of Closure. Some of it is normally avoided because you run your tests on the compiled code too and something breaks because Closure is assuming something you didn't intend, but we obviously don't do that. I'm going to play around with the [closure conformance tests](https://github.com/google/closure-compiler/wiki/JS-Conformance-Framework) they have that can set up stricter checks to see if that helps.

Until then, in general it's good to not use `{Object}` (without a property type parameter) or `{?}` or `{*}` without good cause (or without an associated TODO/issue if it depends on some other code getting types), and to `@struct` ([effectively seals the object](https://github.com/google/closure-compiler/wiki/%40struct-and-%40dict-Annotations)) when we can.

File: aggregators/can-load-offline/index.js (hunk 1 above)
Comment (paullewis) [inline aggregators/can-load-offline/index.js]: Yeah, it complained at me.

File: closure/typedefs/Aggregation.js (hunk 2 above)
Comment (paullewis) [inline closure/typedefs/Aggregation.js]: Done.

File: aggregators/can-load-offline/index.js (hunk 3 above)
Comment (paullewis) [inline aggregators/can-load-offline/index.js]: Done.

File: closure/typedefs/Aggregation.js (hunk 6 above)
Comment (paullewis) [inline closure/typedefs/Aggregation.js]: Done.

File: aggregators/aggregate.js (hunk 4 above)
Comment (paullewis) [inline aggregators/aggregate.js]: Yeah that's DocBlockr. I'll remove the spaces.

File: aggregators/aggregate.js (hunk 5 above)
Comment (paullewis) [inline aggregators/aggregate.js]: Done.

File: closure/typedefs/Aggregation.js (hunk 7 above)
Comment (paullewis) [inline closure/typedefs/Aggregation.js]: Done.

File: closure/typedefs/Aggregation.js (hunk 8 above)
Comment (paullewis) [inline closure/typedefs/Aggregation.js]: Done.

**Comment by paullewis** [comment]
PTAL.

**Comment by brendankenny** [comment]
need to rebase

**Comment by brendankenny** [comment]
(I blame Sam)

**Comment by samccone** [comment]
sorry

**Comment by samccone** [comment]
1 commit also plz

File: aggregators/aggregate.js [hunk 9]
```diff
@@ -89,6 +120,34 @@ class Aggregate {
...
+  static _convertBooleanToWeight(resultValue, expectedValue, weight) {
+    return (resultValue === expectedValue) ? weight : 0;
+  }
+
+  /**
+   * Compares the set of audit results to the expected values.
+   * @param {!Array<!AuditResult>} results The audit results.
+   * @param {!AggregationCriteria} expected The aggregation's expected values and weighting.
+   * @return {AggregationItem} The aggregation score.
```
Comment (brendankenny) [inline aggregators/aggregate.js]: `{!AggregationItem}`

File: closure/typedefs/Aggregation.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ * @struct
+ * @record
+ */
+function Aggregation() {}
+
+/** @type {string} */
+Aggregation.prototype.name;
+
+/** @type {AggregationItem} */
+Aggregation.prototype.score;
```
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: `{!AggregationItem}`

**Comment by brendankenny** [comment]
rebasing might take some work, but this all looks correct to me (with two more `!`s added)

File: closure/typedefs/Aggregation.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+/** @type {(boolean|number|undefined)} */
+AggregationCriterion.prototype.value;
+
+/** @type {number} */
+AggregationCriterion.prototype.weight;
+
+/**
+ * @struct
+ * @type {!Object<string, !AggregationCriterion>}
```
Comment (samthor) [inline closure/typedefs/Aggregation.js]: drive-by nit: This is allowed but needless since string is the default/only key type for an Object. You can just write `!Object<!AggregationCriterion>`.

File: closure/typedefs/Aggregation.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+function AggregationCriterion() {}
+
+/** @type {(boolean|number|undefined)} */
+AggregationCriterion.prototype.value;
+
+/** @type {number} */
+AggregationCriterion.prototype.weight;
+
+/**
+ * @struct
```
Comment (samthor) [inline closure/typedefs/Aggregation.js]: I'm not sure this @struct makes sense here - it implies that it can't be extended but it's an object so it can support arbitrary key/values?

File: aggregators/aggregate.js [hunk 13]
```diff
@@ -54,6 +78,13 @@ class Aggregate {
...
 
+  /**
+   * Converts each raw audit output to a weighted value for the aggregation.
+   * @private
+   * @param {!AuditResult} result The audit's output value.
+   * @param {!AggregationCriterion} expected The aggregation's expected value and weighting for this result.
+   * @return {number} The weighted result.
+   */
   static _convertToWeight(result, expected) {
     let weight = 0;
```
Comment (samthor) [inline aggregators/aggregate.js]: nit on the next line as it relates to types - `typeof expected === 'undefined'` only returns true if undefined is passed, which shouldn't happen because of your Closure annotation. If the aim is to be defensive then a `!expected` check might just be easier.

File: aggregators/aggregate.js [hunk 14]
```diff
@@ -42,6 +61,11 @@ class Aggregate {
     return weight;
   }
 
+  /**
+   * @private
+   * @param {!Array<!AuditResult>} results
+   * @return {!Object}
```
Comment (samthor) [inline aggregators/aggregate.js]: This returns `{!Object<!AuditResult>}` (again, string key is implied).

File: closure/typedefs/Aggregation.js (hunk 12 above)
Comment (brendankenny) [inline closure/typedefs/Aggregation.js]: :+1: more on `@struct` [here](https://github.com/google/closure-compiler/wiki/@struct-and-@dict-Annotations). Basically it's like `Object.seal`, but enforced by the compiler (ignore `@dict`...since we aren't using the compiled output it's not worth it)
